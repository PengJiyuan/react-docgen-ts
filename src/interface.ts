// K-V pair parsed from jsDoc
export type TagType = {
  name: string;
  value: string;
};

// Schema parsed from Symbol
export type PropertyType = {
  name: string;
  type: string;
  isOptional: boolean;
  tags: TagType[];
};

// Schema parsed from function declaration
export type FunctionSchema = {
  tags: TagType[];
  params: Array<PropertyType & { initializerText: string | null }>;
  returns: string;
};

// Schema parsed from interface declaration
export type InterfaceSchema = {
  tags: TagType[];
  data: PropertyType[];
};

// Collect of all schema type generated
export type Schema = FunctionSchema | InterfaceSchema;

// Table type in markdown generated
export type MarkdownTableType = 'interface' | 'parameter';

export type DefaultTypeMapT = Record<string, { type: string; tags: TagType[] }>;

export type GenerateConfig = {
  defaultTypeMap?: DefaultTypeMapT;
  sourceFilesPaths: string | string[];
  /**
   * Whether to skip parsing documentation comment as property description
   */
  strictComment?: boolean;
  /**
   * Generate schema in the order they appear in the document
   * When it's true, generate function will return Array<{ title: string; schema: Schema }>
   */
  strictDeclarationOrder?: boolean;
};

export type GenerateMarkdownConfig = GenerateConfig & {
  lang?: string;
};
