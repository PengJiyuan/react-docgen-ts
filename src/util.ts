export function toSingleLine(str = ''): string {
  if (!str) {
    return '';
  }
  return str
    .trim()
    .replace(/[\r\n\t]+/g, '')
    .replace(/[\x20]{2,}/g, '')
    .replace(/\|/g, '\\|');
}
