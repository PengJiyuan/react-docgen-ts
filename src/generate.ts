import {
  Project,
  SourceFile,
  TypeChecker,
  Symbol,
  FunctionDeclaration,
  InterfaceDeclaration,
  TypeAliasDeclaration,
} from 'ts-morph';
import {
  PropertyType,
  GenerateConfig,
  DefaultTypeMapT,
  TagType,
  FunctionSchema,
  Schema,
} from './interface';
import { defaultTypeMap } from './default';
import { toSingleLine } from './util';

type DeclarationCanBeParsed = InterfaceDeclaration | TypeAliasDeclaration | FunctionDeclaration;

type ExtractType = {
  name: string;
  type: string;
  isOptional: boolean;
};

const TAG_NAMES_FOR_DESCRIPTION = ['zh', 'en'];

const project = new Project({
  compilerOptions: {
    jsx: 'react' as any,
  },
});

const propertyRegex = /(\w+)\s{0,}([?]?)\s{0,}:(.*?);?$/s;

// extract pure type
function extractFromPropertyText(text: string): ExtractType | undefined {
  const regexResult = propertyRegex.exec(text);
  if (!regexResult) {
    return;
  }
  const name = regexResult[1];
  const isOptional = regexResult[2] === '?';
  const type = toSingleLine(regexResult[3]);

  return {
    name,
    isOptional,
    type,
  };
}

// Get key-value pairs from jsDoc of Declarations
function getDeclarationTags(declaration: DeclarationCanBeParsed) {
  const tags: Array<TagType> = [];
  const rawTags = declaration.getJsDocs()[0]?.getTags() || [];
  let title;

  for (const tag of rawTags) {
    const name = tag.getTagName();
    const value = tag.getCommentText() || '';

    if (name === 'title') {
      title = value;
    }

    tags.push({ name, value });
  }

  return {
    title,
    tags,
  };
}

// Get key-value pairs from jsDoc of Symbol
function getSymbolTags(sym: Symbol, strictComment = false): TagType[] {
  const jsDocTags = sym.compilerSymbol.getJsDocTags();
  const tags: TagType[] = jsDocTags.map((tag) => ({
    name: tag.name,
    value: tag.text?.[0].text || '',
  }));

  // Try to extend property description from common comment
  if (!strictComment) {
    const [commonComment] = sym.compilerSymbol.getDocumentationComment(undefined);
    if (commonComment && commonComment.kind === 'text' && commonComment.text) {
      TAG_NAMES_FOR_DESCRIPTION.forEach((tagNameForDescription) => {
        if (!tags.find(({ name }) => name === tagNameForDescription)) {
          tags.push({ name: tagNameForDescription, value: commonComment.text });
        }
      });
    }
  }

  return tags;
}

// Get Json schema of interface's property
function getPropertySchema(
  sym: Symbol,
  defaultT: DefaultTypeMapT,
  strictComment = false
): PropertyType | null {
  const name = sym.getName();
  const typeText = sym.getDeclarations()[0].getText();
  const extract = extractFromPropertyText(typeText);

  if (!extract) {
    return null;
  }

  const tags = getSymbolTags(sym, strictComment);
  if (tags.find(({ name }) => name && TAG_NAMES_FOR_DESCRIPTION.indexOf(name) > -1)) {
    return {
      name,
      type: extract.type,
      isOptional: extract.isOptional,
      tags,
    };
  }

  return defaultT[name]
    ? {
        name,
        isOptional: extract.isOptional,
        ...defaultT[name],
      }
    : null;
}

// Get Json schema of Function
function getFunctionSchema(
  declaration: FunctionDeclaration,
  strictComment = false
): Pick<FunctionSchema, 'params' | 'returns'> {
  return {
    params: declaration.getParameters().map((para) => {
      const tags = getSymbolTags(para.getSymbol() as Symbol, strictComment);
      return {
        tags,
        name: para.getName(),
        type: para.getType().getText(),
        isOptional: para.isOptional(),
        initializerText:
          para.getInitializer()?.getText() ||
          tags.find(({ name }) => name === 'default' || name === 'defaultValue')?.value ||
          null,
      };
    }),
    returns: declaration.getReturnType().getText(),
  };
}

function generateSchema(sourceFile: SourceFile, typeChecker: TypeChecker, config?: GenerateConfig) {
  const interfaces = sourceFile?.getInterfaces() || [];
  const typeAliases = sourceFile?.getTypeAliases() || [];
  const functions = sourceFile?.getFunctions() || [];
  const defaultT = config?.defaultTypeMap || defaultTypeMap;
  const strictComment = !!config?.strictComment;

  const schemaMap: Record<string, Schema> = {};
  const schemaList: Array<{ title: string; schema: Schema }> = [];

  [...interfaces, ...typeAliases, ...functions]
    .sort((declarationA, declarationB) => {
      return declarationA.getStartLineNumber() - declarationB.getStartLineNumber();
    })
    .forEach((declaration) => {
      const { title, tags } = getDeclarationTags(declaration);
      const dType = declaration.getKindName() as
        | 'InterfaceDeclaration'
        | 'FunctionDeclaration'
        | 'TypeAliasDeclaration';

      if (!title) {
        return;
      }

      let schema: Schema;
      const typeNode =
        dType === 'FunctionDeclaration'
          ? (declaration as FunctionDeclaration)
          : dType === 'TypeAliasDeclaration'
          ? (declaration as TypeAliasDeclaration).getTypeNode()
          : null;

      // Function declaration
      if (
        typeNode &&
        ['FunctionDeclaration', 'FunctionType'].indexOf(typeNode.getKindName()) > -1
      ) {
        schema = {
          tags,
          ...getFunctionSchema(typeNode as FunctionDeclaration),
        };
      }
      // Interface declaration forbid extends
      else if (
        dType === 'InterfaceDeclaration' &&
        !!tags.find(({ name }) => name === 'notExtends')
      ) {
        const data: PropertyType[] = [];
        (declaration as InterfaceDeclaration).getProperties().forEach((a) => {
          const schema = getPropertySchema(a.getSymbol() as Symbol, defaultT, strictComment);
          schema && data.push(schema);
        });
        schema = { tags, data };
      } else {
        const data: PropertyType[] = [];
        typeChecker.getPropertiesOfType(declaration.getType()).forEach((a) => {
          const schema = getPropertySchema(a, defaultT, strictComment);
          schema && data.push(schema);
        });
        schema = { tags, data };
      }

      schemaList.push({ title, schema });
      schemaMap[title] = schema;
    });

  return config?.strictDeclarationOrder ? schemaList : schemaMap;
}

function generate(
  file: string,
  config?: GenerateConfig
): Record<string, Schema> | Array<{ title: string; schema: Schema }> | undefined {
  if (config?.sourceFilesPaths) {
    project.addSourceFilesAtPaths(config?.sourceFilesPaths);
  }

  const typeChecker = project.getTypeChecker();
  const sourceFile = project.getSourceFile(file);

  if (!sourceFile) {
    return;
  }

  return generateSchema(sourceFile, typeChecker, config);
}

export default generate;
