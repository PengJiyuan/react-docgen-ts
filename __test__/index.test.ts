import * as path from 'path';
import { generate, generateMarkdown } from '../src/index';

const pathFixtures = path.resolve(__dirname, 'fixtures');
const pathBasic = path.resolve(pathFixtures, 'basic.ts');
const pathExtends = path.resolve(pathFixtures, 'extends/interface.ts');
const pathDefaultMap = path.resolve(pathFixtures, 'defaultTypeMap.ts');
const pathFunction = path.resolve(pathFixtures, 'function.ts');
const pathPropertySorter = path.resolve(pathFixtures, 'propertySorter.ts');

describe('generate', () => {
  it('basic', () => {
    const schema = generate(pathBasic, {
      sourceFilesPaths: './**/*.ts',
    });
    expect(schema).toMatchSnapshot();
  });

  it('extends', () => {
    const schema = generate(pathExtends, {
      sourceFilesPaths: './**/*.ts',
    });
    expect(schema).toMatchSnapshot();
  });

  it('defaultTypeMap', () => {
    const schema = generate(pathDefaultMap, {
      sourceFilesPaths: './**/*.ts',
    });
    expect(schema).toMatchSnapshot();
  });

  it('function type', () => {
    const schema = generate(pathFunction, {
      sourceFilesPaths: './**/*.ts',
    });
    expect(schema).toMatchSnapshot();
  });

  it('sort property', () => {
    const schema = generate(pathPropertySorter, {
      sourceFilesPaths: './**/*.ts',
      propertySorter: ({ type: typeA }, { type: typeB }) => {
        const getLevel = (type) =>
          type === 'boolean'
            ? 0
            : type === 'number'
            ? 1
            : type === 'string'
            ? 2
            : /([^)]*)\s*=>/.test(type)
            ? 3
            : -1;
        return getLevel(typeA) - getLevel(typeB);
      },
    });
    expect(schema).toMatchSnapshot();
  });
});

describe('generateMarkdown', () => {
  it('basic', () => {
    const markdownZh = generateMarkdown(pathBasic, {
      sourceFilesPaths: './**/*.ts',
      lang: 'zh',
    });
    const markdownEn = generateMarkdown(pathBasic, {
      sourceFilesPaths: './**/*.ts',
      lang: 'en',
    });

    expect(markdownZh).toMatchSnapshot();
    expect(markdownEn).toMatchSnapshot();
  });

  it('extends', () => {
    const markdownZh = generateMarkdown(pathExtends, {
      sourceFilesPaths: './**/*.ts',
      lang: 'zh',
    });
    const markdownEn = generateMarkdown(pathExtends, {
      sourceFilesPaths: './**/*.ts',
      lang: 'en',
    });

    expect(markdownZh).toMatchSnapshot();
    expect(markdownEn).toMatchSnapshot();
  });

  it('function type', () => {
    const markdownZh = generateMarkdown(pathFunction, {
      sourceFilesPaths: './**/*.ts',
      lang: 'zh',
    });
    const markdownEn = generateMarkdown(pathFunction, {
      sourceFilesPaths: './**/*.ts',
      lang: 'en',
      strictDeclarationOrder: true,
    });

    expect(markdownZh).toMatchSnapshot();
    expect(markdownEn).toMatchSnapshot();
  });
});
