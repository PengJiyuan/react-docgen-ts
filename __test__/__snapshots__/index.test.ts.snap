// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate basic 1`] = `
Object {
  "A": Object {
    "data": Array [
      Object {
        "isOptional": true,
        "name": "action",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "自定义操作项",
          },
          Object {
            "name": "en",
            "value": "this is action",
          },
          Object {
            "name": "version",
            "value": "2.15.0",
          },
        ],
        "type": "ReactNode",
      },
      Object {
        "isOptional": true,
        "name": "closable",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "是否可以关闭",
          },
          Object {
            "name": "en",
            "value": "Whether Alert can be closed",
          },
          Object {
            "name": "defaultValue",
            "value": "{}",
          },
        ],
        "type": "InnerProps",
      },
    ],
    "tags": Array [
      Object {
        "name": "title",
        "value": "A",
      },
      Object {
        "name": "zh",
        "value": "这是接口描述。",
      },
      Object {
        "name": "en",
        "value": "This is interface description.",
      },
    ],
  },
}
`;

exports[`generate defaultTypeMap 1`] = `
Object {
  "A": Object {
    "data": Array [
      Object {
        "isOptional": false,
        "name": "style",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "节点样式",
          },
          Object {
            "name": "en",
            "value": "Additional style",
          },
        ],
        "type": "CSSProperties",
      },
      Object {
        "isOptional": false,
        "name": "className",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "节点类名",
          },
          Object {
            "name": "en",
            "value": "Additional css class",
          },
        ],
        "type": "string \\\\| string[]",
      },
    ],
    "tags": Array [
      Object {
        "name": "title",
        "value": "A",
      },
    ],
  },
}
`;

exports[`generate extends 1`] = `
Object {
  "Alert": Object {
    "data": Array [
      Object {
        "isOptional": false,
        "name": "action",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "自定义操作项",
          },
          Object {
            "name": "en",
            "value": "this is action",
          },
          Object {
            "name": "version",
            "value": "2.15.0",
          },
        ],
        "type": "ReactElement",
      },
      Object {
        "isOptional": true,
        "name": "closable",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "是否可以关闭",
          },
          Object {
            "name": "en",
            "value": "Whether Alert can be closed",
          },
          Object {
            "name": "defaultValue",
            "value": "false",
          },
        ],
        "type": "{a: boolean;b: string;}",
      },
      Object {
        "isOptional": true,
        "name": "callback",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "回调参数",
          },
          Object {
            "name": "en",
            "value": "Callback function",
          },
        ],
        "type": "(option: Option) => void",
      },
      Object {
        "isOptional": true,
        "name": "bb",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "BB",
          },
          Object {
            "name": "en",
            "value": "bb",
          },
        ],
        "type": "boolean",
      },
      Object {
        "isOptional": true,
        "name": "animation",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "动画",
          },
          Object {
            "name": "en",
            "value": "animation",
          },
        ],
        "type": "boolean \\\\| string",
      },
    ],
    "tags": Array [
      Object {
        "name": "title",
        "value": "Alert",
      },
      Object {
        "name": "zh",
        "value": "向用户显示警告的信息时，通过警告提示，展现需要关注的信息。",
      },
      Object {
        "name": "en",
        "value": "Display warning information to the user. the Alert is used to display the information that needs attention.",
      },
    ],
  },
  "Test": Object {
    "data": Array [
      Object {
        "isOptional": true,
        "name": "position",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "位置",
          },
          Object {
            "name": "en",
            "value": "position",
          },
        ],
        "type": "string",
      },
      Object {
        "isOptional": true,
        "name": "size",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "尺寸",
          },
          Object {
            "name": "en",
            "value": "Size",
          },
        ],
        "type": "string",
      },
      Object {
        "isOptional": false,
        "name": "action",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "自定义操作项",
          },
          Object {
            "name": "en",
            "value": "this is action",
          },
          Object {
            "name": "version",
            "value": "2.15.0",
          },
        ],
        "type": "ReactElement",
      },
      Object {
        "isOptional": true,
        "name": "closable",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "是否可以关闭",
          },
          Object {
            "name": "en",
            "value": "Whether Alert can be closed",
          },
          Object {
            "name": "defaultValue",
            "value": "false",
          },
        ],
        "type": "{a: boolean;b: string;}",
      },
      Object {
        "isOptional": true,
        "name": "callback",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "回调参数",
          },
          Object {
            "name": "en",
            "value": "Callback function",
          },
        ],
        "type": "(option: Option) => void",
      },
      Object {
        "isOptional": true,
        "name": "bb",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "BB",
          },
          Object {
            "name": "en",
            "value": "bb",
          },
        ],
        "type": "boolean",
      },
      Object {
        "isOptional": true,
        "name": "animation",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "动画",
          },
          Object {
            "name": "en",
            "value": "animation",
          },
        ],
        "type": "boolean \\\\| string",
      },
    ],
    "tags": Array [
      Object {
        "name": "title",
        "value": "Test",
      },
    ],
  },
}
`;

exports[`generate function type 1`] = `
Object {
  "Add": Object {
    "params": Array [
      Object {
        "initializerText": null,
        "isOptional": false,
        "name": "a",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "被加数",
          },
          Object {
            "name": "en",
            "value": "First number",
          },
        ],
        "type": "number",
      },
      Object {
        "initializerText": "5",
        "isOptional": true,
        "name": "b",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "加数",
          },
          Object {
            "name": "en",
            "value": "Second number",
          },
        ],
        "type": "number",
      },
    ],
    "returns": "number",
    "tags": Array [
      Object {
        "name": "title",
        "value": "Add",
      },
      Object {
        "name": "en",
        "value": "Add two numbers",
      },
      Object {
        "name": "zh",
        "value": "两数相加",
      },
      Object {
        "name": "returns",
        "value": "Sum of two numbers",
      },
      Object {
        "name": "version",
        "value": "1.0.0",
      },
    ],
  },
  "Button": Object {
    "data": Array [
      Object {
        "isOptional": true,
        "name": "size",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "尺寸",
          },
          Object {
            "name": "en",
            "value": "Size",
          },
          Object {
            "name": "defaultValue",
            "value": "default",
          },
        ],
        "type": "'mini' \\\\| 'large' \\\\| 'default'",
      },
      Object {
        "isOptional": true,
        "name": "color",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "颜色",
          },
          Object {
            "name": "en",
            "value": "Color",
          },
          Object {
            "name": "version",
            "value": "1.2.0",
          },
        ],
        "type": "string",
      },
    ],
    "tags": Array [
      Object {
        "name": "title",
        "value": "Button",
      },
      Object {
        "name": "zh",
        "value": "按钮",
      },
      Object {
        "name": "en",
        "value": "Button",
      },
    ],
  },
  "Plus": Object {
    "params": Array [
      Object {
        "initializerText": null,
        "isOptional": false,
        "name": "a",
        "tags": Array [
          Object {
            "name": "en",
            "value": "First number",
          },
          Object {
            "name": "zh",
            "value": "被乘数",
          },
        ],
        "type": "number",
      },
      Object {
        "initializerText": "1",
        "isOptional": false,
        "name": "b",
        "tags": Array [
          Object {
            "name": "en",
            "value": "Second number",
          },
          Object {
            "name": "zh",
            "value": "乘数",
          },
          Object {
            "name": "defaultValue",
            "value": "1",
          },
        ],
        "type": "number",
      },
    ],
    "returns": "number",
    "tags": Array [
      Object {
        "name": "title",
        "value": "Plus",
      },
      Object {
        "name": "zh",
        "value": "两数相乘",
      },
      Object {
        "name": "en",
        "value": "Multiply two numbers",
      },
      Object {
        "name": "returns",
        "value": "Product of two numbers",
      },
    ],
  },
}
`;

exports[`generate sort property 1`] = `
Object {
  "Button": Object {
    "data": Array [
      Object {
        "isOptional": true,
        "name": "size",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "尺寸",
          },
          Object {
            "name": "en",
            "value": "Size",
          },
          Object {
            "name": "defaultValue",
            "value": "default",
          },
        ],
        "type": "'mini' \\\\| 'large' \\\\| 'default'",
      },
      Object {
        "isOptional": true,
        "name": "readonly",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "只读",
          },
          Object {
            "name": "en",
            "value": "Readonly",
          },
        ],
        "type": "boolean",
      },
      Object {
        "isOptional": true,
        "name": "disabled",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "禁用",
          },
          Object {
            "name": "en",
            "value": "Disabled",
          },
        ],
        "type": "boolean",
      },
      Object {
        "isOptional": true,
        "name": "tabIndex",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "TabIndex",
          },
          Object {
            "name": "en",
            "value": "TabIndex",
          },
        ],
        "type": "number",
      },
      Object {
        "isOptional": true,
        "name": "color",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "颜色",
          },
          Object {
            "name": "en",
            "value": "Color",
          },
          Object {
            "name": "version",
            "value": "1.2.0",
          },
        ],
        "type": "string",
      },
      Object {
        "isOptional": true,
        "name": "renderText",
        "tags": Array [
          Object {
            "name": "zh",
            "value": "自定义文本渲染",
          },
          Object {
            "name": "en",
            "value": "Custom text render",
          },
        ],
        "type": "() => string",
      },
    ],
    "tags": Array [
      Object {
        "name": "title",
        "value": "Button",
      },
      Object {
        "name": "zh",
        "value": "按钮",
      },
      Object {
        "name": "en",
        "value": "Button",
      },
    ],
  },
}
`;

exports[`generateMarkdown basic 1`] = `
Object {
  "A": "### A

这是接口描述。

|参数名|描述|类型|默认值|版本|
|---|---|---|---|---|
|action|自定义操作项|\`ReactNode\`|\`-\`|2.15.0|
|closable|是否可以关闭|\`InnerProps\`|\`{}\`|-|",
}
`;

exports[`generateMarkdown basic 2`] = `
Object {
  "A": "### A

This is interface description.

|Property|Description|Type|DefaultValue|Version|
|---|---|---|---|---|
|action|this is action|\`ReactNode\`|\`-\`|2.15.0|
|closable|Whether Alert can be closed|\`InnerProps\`|\`{}\`|-|",
}
`;

exports[`generateMarkdown custom project 1`] = `
Object {
  "Add": "### Add

两数相加

#### Since
1.0.0

#### Returns
\`number\`: Sum of two numbers

#### Arguments
|参数名|描述|类型|默认值|
|---|---|---|---|
|a|被加数|\`number\` **(必填)**|-|
|b|加数|\`number\`|\`5\`|",
  "Button": "### Button

按钮

|参数名|描述|类型|默认值|版本|
|---|---|---|---|---|
|size|尺寸|\`'mini' \\\\| 'large' \\\\| 'default'\`|\`default\`|-|
|color|颜色|\`string\`|\`-\`|1.2.0|",
  "Plus": "### Plus

两数相乘

#### Returns
\`number\`: Product of two numbers

#### Arguments
|参数名|描述|类型|默认值|
|---|---|---|---|
|a|被乘数|\`number\` **(必填)**|-|
|b|乘数|\`number\` **(必填)**|\`1\`|",
}
`;

exports[`generateMarkdown custom project 2`] = `
Array [
  "### Plus

Multiply two numbers

#### Returns
\`number\`: Product of two numbers

#### Arguments
|Argument|Description|Type|DefaultValue|
|---|---|---|---|
|a|First number|\`number\` **(Required)**|-|
|b|Second number|\`number\` **(Required)**|\`1\`|",
  "### Add

Add two numbers

#### Since
1.0.0

#### Returns
\`number\`: Sum of two numbers

#### Arguments
|Argument|Description|Type|DefaultValue|
|---|---|---|---|
|a|First number|\`number\` **(Required)**|-|
|b|Second number|\`number\`|\`5\`|",
  "### Button

Button

|Property|Description|Type|DefaultValue|Version|
|---|---|---|---|---|
|size|Size|\`'mini' \\\\| 'large' \\\\| 'default'\`|\`default\`|-|
|color|Color|\`string\`|\`-\`|1.2.0|",
]
`;

exports[`generateMarkdown extends 1`] = `
Object {
  "Alert": "### Alert

向用户显示警告的信息时，通过警告提示，展现需要关注的信息。

|参数名|描述|类型|默认值|版本|
|---|---|---|---|---|
|action|自定义操作项|\`ReactElement\` **(必填)**|\`-\`|2.15.0|
|closable|是否可以关闭|\`{a: boolean;b: string;}\`|\`false\`|-|
|callback|回调参数|\`(option: Option) => void\`|\`-\`|-|
|bb|BB|\`boolean\`|\`-\`|-|
|animation|动画|\`boolean \\\\| string\`|\`-\`|-|",
  "Test": "### Test

|参数名|描述|类型|默认值|版本|
|---|---|---|---|---|
|position|位置|\`string\`|\`-\`|-|
|size|尺寸|\`string\`|\`-\`|-|
|action|自定义操作项|\`ReactElement\` **(必填)**|\`-\`|2.15.0|
|closable|是否可以关闭|\`{a: boolean;b: string;}\`|\`false\`|-|
|callback|回调参数|\`(option: Option) => void\`|\`-\`|-|
|bb|BB|\`boolean\`|\`-\`|-|
|animation|动画|\`boolean \\\\| string\`|\`-\`|-|",
}
`;

exports[`generateMarkdown extends 2`] = `
Object {
  "Alert": "### Alert

Display warning information to the user. the Alert is used to display the information that needs attention.

|Property|Description|Type|DefaultValue|Version|
|---|---|---|---|---|
|action|this is action|\`ReactElement\` **(Required)**|\`-\`|2.15.0|
|closable|Whether Alert can be closed|\`{a: boolean;b: string;}\`|\`false\`|-|
|callback|Callback function|\`(option: Option) => void\`|\`-\`|-|
|bb|bb|\`boolean\`|\`-\`|-|
|animation|animation|\`boolean \\\\| string\`|\`-\`|-|",
  "Test": "### Test

|Property|Description|Type|DefaultValue|Version|
|---|---|---|---|---|
|position|position|\`string\`|\`-\`|-|
|size|Size|\`string\`|\`-\`|-|
|action|this is action|\`ReactElement\` **(Required)**|\`-\`|2.15.0|
|closable|Whether Alert can be closed|\`{a: boolean;b: string;}\`|\`false\`|-|
|callback|Callback function|\`(option: Option) => void\`|\`-\`|-|
|bb|bb|\`boolean\`|\`-\`|-|
|animation|animation|\`boolean \\\\| string\`|\`-\`|-|",
}
`;

exports[`generateMarkdown function type 1`] = `
Object {
  "Add": "### Add

两数相加

#### Since
1.0.0

#### Returns
\`number\`: Sum of two numbers

#### Arguments
|参数名|描述|类型|默认值|
|---|---|---|---|
|a|被加数|\`number\` **(必填)**|-|
|b|加数|\`number\`|\`5\`|",
  "Button": "### Button

按钮

|参数名|描述|类型|默认值|版本|
|---|---|---|---|---|
|size|尺寸|\`'mini' \\\\| 'large' \\\\| 'default'\`|\`default\`|-|
|color|颜色|\`string\`|\`-\`|1.2.0|",
  "Plus": "### Plus

两数相乘

#### Returns
\`number\`: Product of two numbers

#### Arguments
|参数名|描述|类型|默认值|
|---|---|---|---|
|a|被乘数|\`number\` **(必填)**|-|
|b|乘数|\`number\` **(必填)**|\`1\`|",
}
`;

exports[`generateMarkdown function type 2`] = `
Array [
  "### Plus

Multiply two numbers

#### Returns
\`number\`: Product of two numbers

#### Arguments
|Argument|Description|Type|DefaultValue|
|---|---|---|---|
|a|First number|\`number\` **(Required)**|-|
|b|Second number|\`number\` **(Required)**|\`1\`|",
  "### Add

Add two numbers

#### Since
1.0.0

#### Returns
\`number\`: Sum of two numbers

#### Arguments
|Argument|Description|Type|DefaultValue|
|---|---|---|---|
|a|First number|\`number\` **(Required)**|-|
|b|Second number|\`number\`|\`5\`|",
  "### Button

Button

|Property|Description|Type|DefaultValue|Version|
|---|---|---|---|---|
|size|Size|\`'mini' \\\\| 'large' \\\\| 'default'\`|\`default\`|-|
|color|Color|\`string\`|\`-\`|1.2.0|",
]
`;
